src/target/riscv/riscv-013.c

```diff
@@ -2811,6 +2818,320 @@ void riscv013_fill_dmi_nop_u64(struct target *target, char *buf)
+			uint32_t wr_data[num_sbdata_regs];
+			for (uint32_t j = 0; j < num_sbdata_regs; j++)
+				wr_data[j] = test_patterns[j] + i;
+			write_memory_sba_simple(target, addr, wr_data, num_sbdata_regs, sbcs);
+		}
+
+		for (uint32_t i = 0; i < num_words; i++) {
+			uint32_t addr = legal_address + (i << sbaccess);
+			read_memory_sba_simple(target, addr, rd_buf, num_sbdata_regs, sbcs);
+			for (uint32_t j = 0; j < num_sbdata_regs; j++) {
+				if (((test_patterns[j]+i)&compare_mask) != (rd_buf[j]&compare_mask)) {
+					LOG_ERROR("System Bus Access Test 1: Error reading non-autoincremented address %x,"
+							"expected val = %x, read val = %x", addr, test_patterns[j]+i, rd_buf[j]);
+					test_passed = false;
+				}
+			}
+		}
+	}
+	if (test_passed)
+		LOG_INFO("System Bus Access Test 1: Simple write/read test PASSED.");
+
+	/* Test 2: Address autoincrement test */
+	target_addr_t curr_addr;
+	target_addr_t prev_addr;
+	test_passed = true;
+	sbcs = set_field(sbcs_orig, DMI_SBCS_SBAUTOINCREMENT, 1);
+	dmi_write(target, DMI_SBCS, sbcs);
+
+	for (uint32_t sbaccess = 0; sbaccess <= (uint32_t)max_sbaccess; sbaccess++) {
+		sbcs = set_field(sbcs, DMI_SBCS_SBACCESS, sbaccess);
+		dmi_write(target, DMI_SBCS, sbcs);
+
+		dmi_write(target, DMI_SBADDRESS0, legal_address);
+		read_sbcs_nonbusy(target, &sbcs);
+		curr_addr = legal_address;
+		for (uint32_t i = 0; i < num_words; i++) {
+			prev_addr = curr_addr;
+			read_sbcs_nonbusy(target, &sbcs);
+			curr_addr = sb_read_address(target);
+			if ((curr_addr - prev_addr != (uint32_t)(1 << sbaccess)) && i != 0) {
+				LOG_ERROR("System Bus Access Test 2: Error with address auto-increment, sbaccess = %x", sbaccess);
+				test_passed = false;
+			}
+			dmi_write(target, DMI_SBDATA0, i);
+		}
+
+		read_sbcs_nonbusy(target, &sbcs);
+
+		dmi_write(target, DMI_SBADDRESS0, legal_address);
+
+		uint32_t val;
+		sbcs = set_field(sbcs, DMI_SBCS_SBREADONDATA, 1);
+		dmi_write(target, DMI_SBCS, sbcs);
+		dmi_read(target, &val, DMI_SBDATA0); // Dummy read to trigger first system bus read
+		curr_addr = legal_address;
+		for (uint32_t i = 0; i < num_words; i++) {
+			prev_addr = curr_addr;
+			read_sbcs_nonbusy(target, &sbcs);
+			curr_addr = sb_read_address(target);
+			if ((curr_addr - prev_addr != (uint32_t)(1 << sbaccess)) && i != 0) {
+				LOG_ERROR("System Bus Access Test 2: Error with address auto-increment, sbaccess = %x", sbaccess);
+				test_passed = false;
+			}
+			dmi_read(target, &val, DMI_SBDATA0);
+			read_sbcs_nonbusy(target, &sbcs);
+			if (i != val) {
+				LOG_ERROR("System Bus Access Test 2: Error reading auto-incremented address,"
+						"expected val = %x, read val = %x", i, val);
+				test_passed = false;
+			}
+		}
+	}
+	if (test_passed)
+		LOG_INFO("System Bus Access Test 2: Address auto-increment test PASSED.");
+
+	/* Test 3: Read from illegal address */
+	read_memory_sba_simple(target, illegal_address, rd_buf, 1, sbcs_orig);
+
+	dmi_read(target, &rd_val, DMI_SBCS);
```

Comment:
if it's not already updated, we should use the new `debug_defines.h` as generated by the spec, so that these error codes match what the spec says